pandas/core/arrays/categorical.py

```diff
@@ -2613,10 +2613,10 @@ def _get_codes_for_values(values, categories):
         # Support inferring the correct extension dtype from an array of
         # scalar objects. e.g.
         # Categorical(array[Period, Period], categories=PeriodIndex(...))
-        try:
-            values = categories.dtype.construct_array_type()._from_sequence(values)
-        except Exception:
-            # but that may fail for any reason, so fall back to object
+        cls = categories.dtype.construct_array_type()
+        values = safe_ea_cast(cls, values)
+        if not isinstance(values, Categorical):
```

Comment:
Is this check correct? This branch is about the `categories` of  Categorical, which are typically not a Categorical themselves.

Reply:
I think you're right, this check should be for `cls`